### ISSUE TITLE
Filter_condition evaluates before query

### ISSUE THREAD

**Issue by willtraweek** [pynamodb/PynamoDB#874]
I'm using a query where I want to sometimes sort by resource_group, resource_name, both or neither (resource_group and resource_name being attributes for the class).  I'm hoping to set queries where a specific resource_group or resource_name can be passed into the function and then a filter_condition set based off of that.  Passing `filter_condition = Event.resource_group.exists()` works just fine, but if I try to pass `filter_condition = (Event.resource_group == resource_group)`, it evaluates to a boolean before the query can read it.  Is there a way to go around this or am I stuck making separate functions for each of the 4 potential options?

Thanks for any and all help in advance!

**Comment by willtraweek** [comment]
At the moment, the code looks like this:
```
def batch_get_events(user_guid: str, limit: int = 100, resource_group: str = None, resource_name: str = None):
    if resource_group and resource_name:
        filter_condition = ((Event.resource_group == resource_group) & (Event.resource_name == resource_name))
    elif resource_group:
        filter_condition = (Event.resource_group == resource_group)
    elif resource_name:
        filter_condition = (Event.resource_name == resource_name)
    else:
        filter_condition = Event.user_guid.exists()

    for item in UserIndex.query(user_guid, limit=limit, filter_condition=filter_condition):
      continue
```

**Comment by navdeep710** [comment]
I don't know whether you have solved it or not, we also faced this problem of premature evaluation, so the problem was our RHS (incoming value) should not be None. Once we ensured that problem resolved

**Comment by willtraweek** [comment]
We ended up "fixing" this by just calling `scan()` and then processing it server side.  Not the best solution but it works for now.

**Comment by ikonst** [comment]
It does sound right that you cannot form an expression by comparing a string attribute to a RHS `None` since that's not a legal value for a string, and does_not_exist is the way to express that.

You can keep the code tidier by bootstrapping with `None` and then chaining with `&=`. This is available since 4.0.0:
https://github.com/pynamodb/PynamoDB/commit/34069dd82321b923ee9e3cebc852ec24440adbba
